book/src/template_expansion.md

```diff
@@ -0,0 +1,387 @@
+::askama::Result::Ok(())
+```
+
+Very much as expected except for the `&(self.x == "a") as &bool`. It allows to
+fallback to a boolean as much as possible, but it also explains why you can't
+do:
+
+```jinja
+{% set x = "a" %}
+{% if x %}
+    {{ x }}
+{% endif %}
+```
+
+Because it fail to compile because:
+
+```console
+error[E0605]: non-primitive cast: `&&str` as `&bool`
+```
+
+### if let
+
+```jinja
+{% if let Some(x) = x %}
+    {{ x }}
+{% endif %}
+```
+
+will generate:
+
+```rust
+if let Some(x) = &(self.x) {
+    writer
+        .write_fmt(
+            format_args!(
+                "{0}",
+                &::askama::MarkupDisplay::new_unsafe(&(x), ::askama::Html),
+            ),
+        )?;
+}
+```
+
+### Loops
+
+```html
+{% for user in users %}
+    {{ user }}
+{% endfor %}
+```
+
+will generate:
+
+```rust
+{
+    let mut _did_loop = false;
+    let _iter = (&self.users).into_iter();
+    for (user, _loop_item) in ::askama::helpers::TemplateLoop::new(_iter) {
+        _did_loop = true;
+        writer
+            .write_fmt(
+                format_args!(
+                    "\n    {0}\n",
+                    &::askama::MarkupDisplay::new_unsafe(&(user), ::askama::Html),
+                ),
+            )?;
+    }
+    if !_did_loop {}
+}
+::askama::Result::Ok(())
+```
+
+Interestingly enough, there is a `_did_loop` variable. It is used in case your
+`for` loop has a `if` condition:
+
+```jinja
+{% for user in x if x.len() > 2 %}
+    {{ user }}
+{% else %}
+    {{ x }}
+{% endfor %}
+```
+
+Which generates:
+
+```rust
+{
+    let mut _did_loop = false;
+    let _iter = (&self.x).into_iter();
+    let _iter = _iter.filter(|user| -> bool { self.x.len() > 2 });
+    for (user, _loop_item) in ::askama::helpers::TemplateLoop::new(_iter) {
+        _did_loop = true;
+        writer
+            .write_fmt(
+                format_args!(
+                    "\n    {0}\n",
+                    &::askama::MarkupDisplay::new_unsafe(&(user), ::askama::Html),
+                ),
+            )?;
+    }
+    if !_did_loop {
+        writer
+            .write_fmt(
+                format_args!(
+                    "\n    {0}\n",
+                    &::askama::MarkupDisplay::new_unsafe(
+                        &(self.x),
+                        ::askama::Html,
+                    ),
+                ),
+            )?;
+    }
+}
+::askama::Result::Ok(())
+```
+
+So in case the code doesn't enter in the loop because of the `if` condition,
+it will run the `else` block.
+
+## Filters
+
+Example by using the `abs` built-in filter:
+
+```jinja
+{{ -2|abs }}
+```
+
+Which generates:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}",
+            &::askama::MarkupDisplay::new_unsafe(
+                &(::askama::filters::abs(-2)?),
+                ::askama::Html,
+            ),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+The filter is called with `-2` as first argument. If you add arguments to the
```

Comment:
The "If" in "If you add" doesn't really lead anywhere? Maybe rewrite as "You can add further arguments to the call like this:".

Reply:
Outch indeed.